zipkin-ui/README.md

```diff
@@ -0,0 +1,34 @@
+# zipkin-ui
+
+Zipkin-UI is a single-page application that reads configuration from `/config.json`.
+
+When looking at a trace, the browser is sent to the path `/traces/{id}`. For the single-page
+app to serve that route, the server needs to forward the request to `index.html`. The same
+forwarding applies to `/dependencies` and any other routes the UI controls. 
+
+Under the scenes the JavaScript code looks at `window.location` to figure out what the
+UI should do. This is handled by a route api defined in the crossroads library.
+
+The full list of
+such routes is defined in [`resources/zipkin-ui/index-routes.json`](src/main/resources/zipkin-ui/index-routes.json).
+Routes ending with `/` should be used in prefix matching (`/foo/` should match `/foo/`, `/foo/bar`, and `/foo/bar/baz`).
+Routes not ending with `/` should be used in exact matching (`/foo` should match `/foo` and nothing else).
+
+Based on all these, there are several ways for serving zipkin-ui:
+
+ * Programatically read `index-routes.json` and set up explicit routes based on it. Example: [zipkin-web](https://github.com/openzipkin/zipkin/blob/5dec252e4c562b21bac5ac2f9d0b437d90988f79/zipkin-web/src/main/scala/com/twitter/zipkin/web/Main.scala#L92-L95).
+   **Pro**: serves `index.html` for exactly the right routes. **Con**: needs active server component. **Suggested** method when
+   embedding in a server application.
+ * Match all unknown paths, and instead of returning 404, route them back to `index.html`. Example: [zipkin-query](https://github.com/openzipkin/zipkin/blob/5dec252e4c562b21bac5ac2f9d0b437d90988f79/zipkin-query/src/main/scala/com/twitter/zipkin/query/ZipkinQueryController.scala).
+   **Pro**: never need to touch the routing rules again. **Con**: no 404, may interfere with other route definitions.
+   **Suggested** method for static webservers.
+ * A variation on the above: only serve `index.html` for paths where the last `/`-separated
+   segment doesn't have a `.` in it. **Pros**: never need to touch the routing rules again; has 404 for paths with an extension.
+   **Con**: the heuristic is not very robust.
+ * Manually hard-code all the routes that need to serve `index.html`. Example: [zipkin-java](https://github.com/openzipkin/zipkin-java/blob/d031021769432f81d60690356249fee56003c5d1/zipkin-server/src/main/java/zipkin/server/ZipkinUI.java)
+   **Pro**: doesn't interfere with other route definitions, correct 404s. **Con**: needs updating when new UI endpoints are
+   added to Zipkin-UI.
+
+## What? Why? Why not a .* package?
```

Comment:
"Why is it packaged as a .jar and not a .zip?"
Which could be answered by something like "since many zipkin servers are java-based, it's convenient to distribute the ui as a jar, which can be imported by the Gradle build tool. A .jar file is really only a .zip file, and can be treated as such. It can be opened by any program that can extract zip files."